{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/dask/dataframe/__init__.py:42: FutureWarning: \n",
      "Dask dataframe query planning is disabled because dask-expr is not installed.\n",
      "\n",
      "You can install it with `pip install dask[dataframe]` or `conda install dask`.\n",
      "This will raise in a future version.\n",
      "\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "import dask.dataframe as dd\n",
    "\n",
    "# Initialize Dask (optional)\n",
    "from dask.distributed import Client\n",
    "client = Client()  # This will start a local cluster\n",
    "\n",
    "# Define the path to your Parquet file\n",
    "parquet_file = 'data/itineraries_gzip.parquet'\n",
    "\n",
    "# Read the Parquet file with Dask\n",
    "df = dd.read_parquet(parquet_file)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/dask/dataframe/__init__.py:42: FutureWarning: \n",
      "Dask dataframe query planning is disabled because dask-expr is not installed.\n",
      "\n",
      "You can install it with `pip install dask[dataframe]` or `conda install dask`.\n",
      "This will raise in a future version.\n",
      "\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/dask/dataframe/__init__.py:42: FutureWarning: \n",
      "Dask dataframe query planning is disabled because dask-expr is not installed.\n",
      "\n",
      "You can install it with `pip install dask[dataframe]` or `conda install dask`.\n",
      "This will raise in a future version.\n",
      "\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/dask/dataframe/__init__.py:42: FutureWarning: \n",
      "Dask dataframe query planning is disabled because dask-expr is not installed.\n",
      "\n",
      "You can install it with `pip install dask[dataframe]` or `conda install dask`.\n",
      "This will raise in a future version.\n",
      "\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/dask/dataframe/__init__.py:42: FutureWarning: \n",
      "Dask dataframe query planning is disabled because dask-expr is not installed.\n",
      "\n",
      "You can install it with `pip install dask[dataframe]` or `conda install dask`.\n",
      "This will raise in a future version.\n",
      "\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "2024-10-18 11:09:11,064 - distributed.worker - ERROR - failed during get data with tcp://127.0.0.1:55729 -> tcp://127.0.0.1:55731\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/tornado/iostream.py\", line 962, in _handle_write\n",
      "    num_bytes = self.write_to_fd(self._write_buffer.peek(size))\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/tornado/iostream.py\", line 1124, in write_to_fd\n",
      "    return self.socket.send(data)  # type: ignore\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [Errno 55] No buffer space available\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/worker.py\", line 1778, in get_data\n",
      "    response = await comm.read(deserializers=serializers)\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/comm/tcp.py\", line 236, in read\n",
      "    convert_stream_closed_error(self, e)\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/comm/tcp.py\", line 140, in convert_stream_closed_error\n",
      "    raise CommClosedError(f\"in {obj}: {exc.__class__.__name__}: {exc}\") from exc\n",
      "distributed.comm.core.CommClosedError: in <TCP (closed)  local=tcp://127.0.0.1:55729 remote=tcp://127.0.0.1:55888>: OSError: [Errno 55] No buffer space available\n",
      "2024-10-18 11:09:16,499 - distributed.worker - ERROR - Worker stream died during communication: tcp://127.0.0.1:55729\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/comm/tcp.py\", line 227, in read\n",
      "    frames_nosplit = await read_bytes_rw(stream, frames_nosplit_nbytes)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/comm/tcp.py\", line 366, in read_bytes_rw\n",
      "    actual = await stream.read_into(chunk)  # type: ignore[arg-type]\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "tornado.iostream.StreamClosedError: Stream is closed\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/worker.py\", line 2056, in gather_dep\n",
      "    response = await get_data_from_worker(\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/worker.py\", line 2874, in get_data_from_worker\n",
      "    response = await send_recv(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/core.py\", line 1015, in send_recv\n",
      "    response = await comm.read(deserializers=deserializers)\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/comm/tcp.py\", line 236, in read\n",
      "    convert_stream_closed_error(self, e)\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/comm/tcp.py\", line 142, in convert_stream_closed_error\n",
      "    raise CommClosedError(f\"in {obj}: {exc}\") from exc\n",
      "distributed.comm.core.CommClosedError: in <TCP (closed) Ephemeral Worker->Worker for gather local=tcp://127.0.0.1:55888 remote=tcp://127.0.0.1:55729>: Stream is closed\n",
      "2024-10-18 11:10:49,741 - distributed.worker.memory - WARNING - Unmanaged memory use is high. This may indicate a memory leak or the memory may not be released to the OS; see https://distributed.dask.org/en/latest/worker-memory.html#memory-not-released-back-to-the-os for more information. -- Unmanaged memory: 3.33 GiB -- Worker memory limit: 4.50 GiB\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125177' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125157' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125179' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125229' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125168' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125166' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125171' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125162' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125159' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125156' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125163' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125191' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125233' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125173' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125234' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125216' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125209' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125225' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125227' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125230' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125197' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125175' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125180' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125223' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125210' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125212' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125188' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125184' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125208' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125206' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-125205' coro=<Client._gather.<locals>.wait() done, defined at /Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py:2385> exception=AllExit()>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samrichards/miniconda3/envs/pydev/lib/python3.11/site-packages/distributed/client.py\", line 2394, in wait\n",
      "    raise AllExit()\n",
      "distributed.client.AllExit\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Save each of them to separate CSV files\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m legId \u001b[38;5;129;01min\u001b[39;00m legIds:\n\u001b[0;32m----> 6\u001b[0m     \u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mlegId\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mlegId\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompute\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mto_csv(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdata/legs/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlegId\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m'\u001b[39m, index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "File \u001b[0;32m~/miniconda3/envs/pydev/lib/python3.11/site-packages/dask/base.py:372\u001b[0m, in \u001b[0;36mDaskMethodsMixin.compute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    348\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    349\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Compute this dask collection\u001b[39;00m\n\u001b[1;32m    350\u001b[0m \n\u001b[1;32m    351\u001b[0m \u001b[38;5;124;03m    This turns a lazy Dask collection into its in-memory equivalent.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    370\u001b[0m \u001b[38;5;124;03m    dask.compute\u001b[39;00m\n\u001b[1;32m    371\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 372\u001b[0m     (result,) \u001b[38;5;241m=\u001b[39m \u001b[43mcompute\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtraverse\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    373\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/miniconda3/envs/pydev/lib/python3.11/site-packages/dask/base.py:660\u001b[0m, in \u001b[0;36mcompute\u001b[0;34m(traverse, optimize_graph, scheduler, get, *args, **kwargs)\u001b[0m\n\u001b[1;32m    657\u001b[0m     postcomputes\u001b[38;5;241m.\u001b[39mappend(x\u001b[38;5;241m.\u001b[39m__dask_postcompute__())\n\u001b[1;32m    659\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m shorten_traceback():\n\u001b[0;32m--> 660\u001b[0m     results \u001b[38;5;241m=\u001b[39m \u001b[43mschedule\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdsk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    662\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m repack([f(r, \u001b[38;5;241m*\u001b[39ma) \u001b[38;5;28;01mfor\u001b[39;00m r, (f, a) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(results, postcomputes)])\n",
      "File \u001b[0;32m~/miniconda3/envs/pydev/lib/python3.11/threading.py:629\u001b[0m, in \u001b[0;36mEvent.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    627\u001b[0m signaled \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_flag\n\u001b[1;32m    628\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m signaled:\n\u001b[0;32m--> 629\u001b[0m     signaled \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cond\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    630\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m signaled\n",
      "File \u001b[0;32m~/miniconda3/envs/pydev/lib/python3.11/threading.py:331\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    329\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    330\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 331\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m \u001b[43mwaiter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    332\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    333\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m waiter\u001b[38;5;241m.\u001b[39macquire(\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Take the 10 legIds with the most rows\n",
    "legIds = df['legId'].value_counts().compute().head(10).index.tolist()\n",
    "\n",
    "# Save each of them to separate CSV files\n",
    "for legId in legIds:\n",
    "    df[df['legId'] == legId].compute().to_csv(f'data/legs/{legId}.csv', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "legs = [\"318ae5fe46ac6a4ed5e1ee55f5b53677\", \"6365830250822d8732e98ac28735f7b0\", \"65f59b3c81d5fcdea844c33e88cb2348\", \"809668b03345672b1487b8582c0b3fd2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-04 10:44:22,134 - distributed.worker.memory - WARNING - Worker is at 83% memory usage. Pausing worker.  Process memory: 3.75 GiB -- Worker memory limit: 4.50 GiB\n",
      "2024-10-04 10:44:22,205 - distributed.worker.memory - WARNING - Worker is at 66% memory usage. Resuming worker. Process memory: 3.00 GiB -- Worker memory limit: 4.50 GiB\n"
     ]
    }
   ],
   "source": [
    "# Get the rows for this 318ae5fe46ac6a4ed5e1ee55f5b53677 legId and save to a new CSV file\n",
    "df[df['legId'] == '318ae5fe46ac6a4ed5e1ee55f5b53677'].compute().to_csv('data/legs/318ae5fe46ac6a4ed5e1ee55f5b53677.csv', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seatsRemaining\n",
      "7     28722589\n",
      "9     21850978\n",
      "0      5337946\n",
      "1      4876317\n",
      "4      4815097\n",
      "2      4553753\n",
      "3      4135191\n",
      "5      3281851\n",
      "6      2966026\n",
      "8      1337697\n",
      "10      261308\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Get counts of seatsRemaining for each legId\n",
    "print(df['seatsRemaining'].value_counts().compute())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pydev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
